### PR TITLE
Fix bug: kubelet failed to clear corrupted subpath dir

### LINKED ISSUE #68316: Kubelet failed to clear subpath dir if the backend NFS volume is not available
/kind bug

**What happened**:
Kubelet failed to clear subpath dir if the backend NFS volume is not available. 
**What you expected to happen**:
When the backend NFS volume is not available, the subpath can also be successfully cleared.

**How to reproduce it (as minimally and precisely as possible)**:
Mount an network volume and disconnect/de-provision target.
```
    spec:
      volumes:
        - name: pvc-153619852761014261
          persistentVolumeClaim:
            claimName: pvc-153619852761014261
      containers:
        - name: container-0
          image: 'nginx:latest'
          volumeMounts:
            - name: pvc-153619852761014261
              mountPath: /nfsvolume
              subPath: nginx
```

**Anything else we need to know?**:
E0905 17:24:54.698586   75516 nestedpendingoperations.go:263] Operation for "\"flexvolume-huawei.com/fuxinfs/280e83b2-b0ed-11e8-b1a9-fa163e6ae1b1-pvc-282497ed-b0ed-11e8-b1a9-fa163e6ae1b1\" (\"280e83b2-b0ed-11e8-b1a9-fa163e6ae1b1\")" failed. No retries permitted until 2018-09-05 17:24:55.19850185 +0800 CST m=+846.029487236 (durationBeforeRetry 500ms). Error: "error cleaning subPath mounts for volume \"pvc-153613928346479792\" (UniqueName: \"flexvolume-huawei.com/fuxinfs/280e83b2-b0ed-11e8-b1a9-fa163e6ae1b1-pvc-282497ed-b0ed-11e8-b1a9-fa163e6ae1b1\") pod \"280e83b2-b0ed-11e8-b1a9-fa163e6ae1b1\" (UID: \"280e83b2-b0ed-11e8-b1a9-fa163e6ae1b1\") : error reading /var/lib/kubelet/pods/280e83b2-b0ed-11e8-b1a9-fa163e6ae1b1/volume-subpaths/pvc-282497ed-b0ed-11e8-b1a9-fa163e6ae1b1/container-0: lstat /var/lib/kubelet/pods/280e83b2-b0ed-11e8-b1a9-fa163e6ae1b1/volume-subpaths/pvc-282497ed-b0ed-11e8-b1a9-fa163e6ae1b1/container-0/0: stale NFS file handle"

**Environment**:
- Kubernetes version (use `kubectl version`): 1.9.7
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):EulerOS 2.0    
- Kernel (e.g. `uname -a`):3.10.0-327.62.59.83.h96.x86_64
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by linyouchong**
**What this PR does / why we need it**:
Fix bug: kubelet failed to clear corrupted subpath dir

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68316

**Special notes for your reviewer**:
NONE

**Release note**:
```release-note
Fix bug: kubelet failed to clear corrupted subpath dir
```

### CODE REVIEW

File: pkg/util/mount/mount_linux.go [hunk 1]
```diff
@@ -892,7 +892,14 @@ func doCleanSubPaths(mounter Interface, podDir string, volumeName string) error
 		fullContainerDirPath := filepath.Join(subPathDir, containerDir.Name())
 		subPaths, err := ioutil.ReadDir(fullContainerDirPath)
 		if err != nil {
-			return fmt.Errorf("error reading %s: %s", fullContainerDirPath, err)
+			glog.Errorf("error reading %s: %s", fullContainerDirPath, err)
+			if isCorruptedMnt(err) {
+				for i := 0; pathExists(path.Join(fullContainerDirPath, fmt.Sprint(i))); i++ {
```
Comment (andyzhangx) [inline pkg/util/mount/mount_linux.go]: could you add comment here? why would you loop from 0 to n here? what's the expected count usually?

File: pkg/util/mount/mount_linux.go [hunk 2]
```diff
@@ -944,6 +951,57 @@ func doCleanSubPath(mounter Interface, fullContainerDirPath, subPathIndex string
 	return nil
 }
 
+// cleanCorruptedMnt tears down the corrupted subpath bind mount
+func cleanCorruptedMnt(mounter Interface, fullSubPath string) error {
```
Comment (andyzhangx) [inline pkg/util/mount/mount_linux.go]: add unit test for this?

File: pkg/util/mount/mount_linux.go [hunk 3]
```diff
@@ -944,6 +951,57 @@ func doCleanSubPath(mounter Interface, fullContainerDirPath, subPathIndex string
...
+	// Remove it *non*-recursively, just in case there were some hiccups.
+	if err := os.Remove(fullSubPath); err != nil {
+		return fmt.Errorf("error deleting %s: %s", fullSubPath, err)
+	}
+	glog.V(5).Infof("Removed %s", fullSubPath)
+	return nil
+}
+
+// pathExists returns true if the specified path exists.
+func pathExists(path string) bool {
```
Comment (andyzhangx) [inline pkg/util/mount/mount_linux.go]: there is already `volumeutil.PathExists` func, could you use that?

File: pkg/util/mount/mount_linux.go [hunk 4]
```diff
@@ -944,6 +951,57 @@ func doCleanSubPath(mounter Interface, fullContainerDirPath, subPathIndex string
...
+		return false
+	} else if isCorruptedMnt(err) {
+		return true
+	} else {
+		return false
+	}
+}
+
+// isCorruptedMnt return true if err is about corrupted mount point
+func isCorruptedMnt(err error) bool {
```
Comment (andyzhangx) [inline pkg/util/mount/mount_linux.go]: There is `IsCorruptedMnt` func under pkg/volume/util/util.go, could you use that?

File: pkg/util/mount/mount_linux.go (hunk 3 above)
Comment (linyouchong) [reply pkg/util/mount/mount_linux.go]: I have tried, but the compiler complaints `import cycle not allowed`

Comment (linyouchong) [reply pkg/util/mount/mount_linux.go]: I have tried, but the compiler complaints `import cycle not allowed`

File: pkg/util/mount/mount_linux.go (hunk 4 above)
Comment (linyouchong) [reply pkg/util/mount/mount_linux.go]: I have tried, but the compiler complaints `import cycle not allowed`

File: pkg/util/mount/mount_linux.go (hunk 2 above)
Comment (linyouchong) [reply pkg/util/mount/mount_linux.go]: Just `umount` and `remove` operation, it almost the same with `doCleanSubPath`

File: pkg/util/mount/mount_linux.go (hunk 1 above)
Comment (linyouchong) [reply pkg/util/mount/mount_linux.go]: Thanks for your review.
Comment added!

**Comment by matti** [comment]
Don't let this rot - every AWS EFS setup bugs with this

**Comment by jingxu97** [comment]
cc @msau42

**Comment by jingxu97** [comment]
is this PR trying to solve the same problem as PR https://github.com/kubernetes/kubernetes/pull/71804?

**Comment by dtaniwaki** [comment]
I found this problem in my production environment and need the fix. Is someone still trying to handle it?

**Comment by msau42** [comment]
This issue is fixed by https://github.com/kubernetes/kubernetes/pull/71804

**Pull request closed without merging**